dstep/translator/Translator.d

```diff
@@ -69,48 +70,62 @@ class Translator
         write(outputFile, translateToString());
     }
 
+    static size_t countTrailingEndls(string a) 
+    {
+        size_t itr = a.length;
+
+        while (itr != 0 && a[itr - 1] == '\n')
+            --itr;
+                
+        return a.length - itr;
+    }
+
+    static size_t countLeadingEndls(string a)
+    {
+        size_t count = 0;
+
+        while (count < a.length && a[count] == '\n')
+            ++count;
+
+        return count;
+    }
+
+    static void endlAwareAppend(ref Appender!string a, string b)
+    {
+        import std.algorithm.comparison;
+
+        size_t aTrailing = countTrailingEndls(a.data);
+        size_t bLeading = countLeadingEndls(b);
+        size_t bTrailing = countTrailingEndls(b);
+
+        size_t num = min(bLeading + aTrailing, 2);
+        a.put("\n".replicate(num < aTrailing ? 0 : num - aTrailing));
+        a.put(b[bLeading..$-bTrailing]);
+        a.put("\n".replicate(min(bTrailing, 2)));
+    }
```

Comment:
So, maybe instead of returning strings from translation subroutines we could return some abstract type like `Node`:

    struct Node
    {
        string spelling;
        bool endlHint; // tells if node wants space around itself
        Node[] children;
    }

Then during the translation process we would create a tree of such a nodes and then there could be a separate routine that takes a root of such a tree and efficiently produces output string. We could even handle making indentations there.

`endlHint` could be extended to something more than `bool` e.g. `enum`:

    enum EndlHint
    {
        Inline, // no newline at all
        NewLine, // should occupy separate line
        Group, // should has blank line before and after
        ...
    }